### PR TITLE
[ntuple] allow RNTuple to emulate fields of unknown types via RecordFields

### PR DESCRIPTION
**Pull request by silverweed**
# This Pull request:
adds the option to "emulate" fields of unknown user-defined types (i.e. types for which we lack dictionaries) as RecordFields whose structure is reconstructed from the on-disk information. This is off by default but can be toggled via `RCreateModelOptions`.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

### CODE REVIEW

**Comment by silverweed** [comment]
Unfortunately this approach doesn't work in all cases.

When replacing the InvalidFields we would also need to patch up stuff like fItemSize on the parent, which is brittle to do.
Probably the only robust way to go is to construct the field hierarchy bottom-up. To do so, my and Jakob's idea is:
- change RFieldBase::Create() so that it returns a "FieldGenerator" rather than a field. This is a functor that accepts a Descriptor for the field and returns a RFieldBase
- recursively construct a tree of Generators from RFieldBase::Create()
- from RFieldDescriptor::CreateField(), visit the tree of generators and replace all RInvalidField generators with the respective record field generators, then call the generators depth-first to construct the field tree.

**Comment by pcanal** [comment]
I am confused.  Why do we need to emulate the case?  What is the usage pattern we are trying to emulated?  Isn't it enough to write the `RNTuple` in one process with the dictionary and read it in a different process without the dictionary?

**Comment by silverweed** [comment]
@pcanal The use case is reading a RNTuple as best as possible when the dictionaries for some of the types stored are not available. For example we still want to be able to print out the schema of the RNTuple even when we have no clue about the specifics of the types stored.
Maybe the word "emulate" is not the best choice here, but the idea is adding the capability for RNTuple to reconstruct the type hierarchy without actually recreating the original classes, but "emulating" them with Record fields instead.

File: tree/ntuple/v7/test/ntuple_fork.hxx [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#pragma once
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_fork.hxx:1]: We generally don't use `#pragma once` in ROOT and RNTuple

File: tree/ntuple/v7/test/ntuple_evolution.cxx [hunk 2]
```diff
@@ -1,62 +1,14 @@
 #include "ntuple_test.hxx"
 
+#define NTUPLE_FORK_MACROS
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_evolution.cxx:3]: Eww, this is ugly. Can we just have all implementations in a header file that we can include into multiple tests?

File: tree/ntuple/v7/test/ntuple_fork.hxx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+// These helpers are split into an *Impl function and a macro to check for and propagate fatal failures to the caller.
+
+void ExecInForkImpl(std::function<void(void)> fn);
+
+void DeclarePointerImpl(std::string_view type, std::string_view variable, void *ptr);
+
+void ProcessLineImpl(const char *line);
+
+#ifdef NTUPLE_FORK_MACROS
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_fork.hxx:14]: Why do we need to protect these macro definitions? But see also my other comment if we can just have all implementations in a header file

File: tree/ntuple/v7/test/ntuple_evolution.cxx [hunk 4]
```diff
@@ -32,7 +32,7 @@ TEST(RNTupleEvolution, AddedMember)
 {
    FileRaii fileGuard("test_ntuple_evolution_added_member.root");
 
-   WriteOldInFork([&] {
+   ExecInFork([&] {
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_evolution.cxx:35]: These changes have to go into the commit that renamed the functions!

File: tree/ntuple/v7/test/ntuple_emulated.cxx [hunk 5]
```diff
@@ -61,7 +61,7 @@ struct Outer {
       try {
          auto model = desc.CreateModel(opts);
          FAIL() << "Creating a model without fEmulateUnknownTypes should fail";
-      } catch (RException ex) {
+      } catch (ROOT::RException ex) {
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_emulated.cxx:64]: has to go into the commit that added the test

File: tree/ntuple/v7/test/ntuple_evolution.cxx [hunk 6]
```diff
@@ -579,7 +579,7 @@ TEST(RNTupleEvolution, AddedIntermediateClass)
 {
    FileRaii fileGuard("test_ntuple_evolution_added_intermediate_class.root");
 
-   WriteOldInFork([&] {
+   ExecInFork([&] {
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_evolution.cxx:582]: has to go into the commit that renamed the functions

File: tree/ntuple/v7/inc/ROOT/RFieldBase.hxx [hunk 7]
```diff
@@ -473,9 +481,15 @@ protected:
    /// normalized type name and type alias
    /// TODO(jalopezg): this overload may eventually be removed leaving only the `RFieldBase::Create()` that takes a
    /// single type name
-   static RResult<std::unique_ptr<RFieldBase>> Create(const std::string &fieldName, const std::string &canonicalType,
-                                                      const std::string &typeAlias, bool continueOnError = false,
-                                                      bool emulateMissingTypes = false);
+   static RResult<std::unique_ptr<RFieldBase>>
+   Create(const std::string &fieldName, const std::string &canonicalType, const std::string &typeAlias,
+          const RCreateFieldOptions &options = {}, const RNTupleDescriptor *desc = nullptr,
+          DescriptorId_t fieldId = kTraitInvalidField);
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/RFieldBase.hxx:487]: This looks odd to me: Is `kTraitInvalidField` really (semantically) a `DescriptorId_t`?

File: tree/ntuple/v7/inc/ROOT/RFieldBase.hxx [hunk 8]
```diff
@@ -244,6 +245,13 @@ private:
       }
    }
 
+   void SetEmulatedFromType(std::string_view type)
+   {
+      assert(fType.empty());
+      fTraits |= kTraitEmulatedField;
+      fType = type;
+   }
+
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/RFieldBase.hxx:254]: Do we need this generically in `RFieldBase` or can this move to `RRecordField`? potentially as a constructor option or with a factory method (was we have `RVectorField::CreateUntyped()`)

File: tree/ntuple/v7/src/RFieldSequenceContainer.cxx [hunk 9]
```diff
@@ -586,6 +586,7 @@ void ROOT::Experimental::RVectorField::RVectorDeleter::operator()(void *objPtr,
 {
    auto vecPtr = static_cast<std::vector<char> *>(objPtr);
    if (fItemDeleter) {
+      R__ASSERT(fItemSize > 0);
```
Comment (hahnjo) [inline tree/ntuple/v7/src/RFieldSequenceContainer.cxx:589]: This looks unrelated to the emulation itself

File: tree/ntuple/v7/test/ntuple_emulated.cxx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+   auto reader = RNTupleReader::Open("ntpl", fileGuard.GetPath());
+   const auto &desc = reader->GetDescriptor();
+
+   {
+      RNTupleDescriptor::RCreateModelOptions opts;
+      opts.fEmulateUnknownTypes = false;
+      try {
+         auto model = desc.CreateModel(opts);
+         FAIL() << "Creating a model without fEmulateUnknownTypes should fail";
+      } catch (RException ex) {
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_emulated.cxx]: change to `ROOT::RException` needs to move here from another commit, and the compiler complains:
```
catching polymorphic type ‘class ROOT::RException’ by value
```

File: tree/ntuple/v7/test/ntuple_emulated.cxx [hunk 11]
```diff
@@ -94,3 +94,107 @@ struct Outer {
...
+   auto reader = RNTupleReader::Open("ntpl", fileGuard.GetPath());
+   const auto &desc = reader->GetDescriptor();
+
+   {
+      RNTupleDescriptor::RCreateModelOptions opts;
+      opts.fEmulateUnknownTypes = false;
+      try {
+         auto model = desc.CreateModel(opts);
+         FAIL() << "Creating a model without fEmulateUnknownTypes should fail";
+      } catch (ROOT::RException ex) {
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_emulated.cxx:156]: The compiler complains
```
catching polymorphic type ‘class ROOT::RException’ by value
```

**Review by hahnjo — CHANGES_REQUESTED** [review]
I left some initial comments. In general, I find this PR hard to follow because some implementations get changed multiple times and some commits introduce options but only half implement them. Can you maybe try to clean this up a bit? Also many of the intermediate commits will probably not compile / pass the tests...

File: tree/ntuple/v7/inc/ROOT/RFieldBase.hxx (hunk 7 above)
Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RFieldBase.hxx:487]: No, this was an oversight: it was supposed to be kInvalidDescriptorId. Thanks for spotting it.

File: tree/ntuple/v7/src/RFieldSequenceContainer.cxx (hunk 9 above)
Comment (silverweed) [reply tree/ntuple/v7/src/RFieldSequenceContainer.cxx:589]: I found a crash while testing it and it didn't look like it was worth a separate PR (or even commit)...

File: tree/ntuple/v7/inc/ROOT/RFieldBase.hxx [hunk 12]
```diff
@@ -460,8 +474,14 @@ protected:
...
-   static RResult<std::unique_ptr<RFieldBase>> Create(const std::string &fieldName, const std::string &canonicalType,
-                                                      const std::string &typeAlias, bool continueOnError = false);
+   static RResult<std::unique_ptr<RFieldBase>>
+   Create(const std::string &fieldName, const std::string &canonicalType, const std::string &typeAlias,
+          const RCreateFieldOptions &options = {}, const RNTupleDescriptor *desc = nullptr,
+          DescriptorId_t fieldId = kInvalidDescriptorId);
+
+   static RResult<std::unique_ptr<RFieldBase>> Create(const std::string &fieldName, const std::string &typeAlias,
+                                                      const RCreateFieldOptions &options, const RNTupleDescriptor *desc,
+                                                      DescriptorId_t fieldId);
```
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RFieldBase.hxx:484]: I think we should update the method documentation for the new parameters

File: tree/ntuple/v7/src/RFieldBase.cxx [hunk 13]
```diff
@@ -278,32 +289,40 @@ ROOT::Experimental::RFieldBase::Check(const std::string &fieldName, const std::s
...
+   RCreateFieldOptions cfOpts{};
+   cfOpts.fReturnInvalidOnError = true;
+   cfOpts.fEmulateUnknownTypes = false;
+   fieldZero.Attach(RFieldBase::Create(fieldName, canonicalType, typeAlias, cfOpts).Unwrap());
 
    std::vector<RCheckResult> result;
    for (const auto &f : fieldZero) {
-      auto invalidField = dynamic_cast<const RInvalidField *>(&f);
-      if (!invalidField)
+      bool isInvalidField = f.GetTraits() & RFieldBase::kTraitInvalidField;
```
Comment (jblomer) [inline tree/ntuple/v7/src/RFieldBase.cxx:299]:
```suggestion
      const bool isInvalidField = f.GetTraits() & RFieldBase::kTraitInvalidField;
```

File: tree/ntuple/v7/src/RFieldBase.cxx [hunk 14]
```diff
@@ -314,11 +333,20 @@ ROOT::Experimental::RFieldBase::Create(const std::string &fieldName, const std::
 
    std::unique_ptr<ROOT::Experimental::RFieldBase> result;
 
+   const auto maybeGetChildId = [desc, fieldId](int childIdx) {
+      if (desc) {
+         const auto &fieldDesc = desc->GetFieldDescriptor(fieldId);
+         return fieldDesc.GetLinkIds()[childIdx];
```
Comment (jblomer) [inline tree/ntuple/v7/src/RFieldBase.cxx:339]: Let's access with bounds check, just in case the descriptor is corrupted.
```suggestion
         return fieldDesc.GetLinkIds().at(childIdx);
```

File: tree/ntuple/v7/src/RFieldBase.cxx [hunk 15]
```diff
@@ -560,13 +599,28 @@ ROOT::Experimental::RFieldBase::Create(const std::string &fieldName, const std::
                   result = std::make_unique<RClassField>(fieldName, canonicalType);
                }
             }
+         } else if (options.fEmulateUnknownTypes) {
+            assert(desc);
```
Comment (jblomer) [inline tree/ntuple/v7/src/RFieldBase.cxx:603]: This option can be set through the public API and then we don't have a descriptor, no?

**Review by jblomer — COMMENTED** [review]
Very nice! We should also add and test emulated collection proxies (probably in another PR).

File: tree/ntuple/v7/test/ntuple_evolution.cxx (hunk 6 above)
Comment (hahnjo) [reply tree/ntuple/v7/test/ntuple_evolution.cxx:581]: still needs to go into the commit that renamed the function

Comment (silverweed) [reply tree/ntuple/v7/test/ntuple_evolution.cxx:581]: I was sure I moved this, no idea how did this slip through.

File: tree/ntuple/v7/src/RFieldBase.cxx (hunk 15 above)
Comment (silverweed) [reply tree/ntuple/v7/src/RFieldBase.cxx:610]: With [e2c3074](https://github.com/root-project/root/pull/17288/commits/e2c3074bf4a2a3af219060f45f24ceb2da346dfa) the user cannot directly call `RFieldBase::Create()` with `fEmulateUnknownTypes == true`, so this code path is only taken when called by `RFieldDescriptor::CreateField` (which does pass down the descriptor)

File: tree/ntuple/v7/src/RFieldBase.cxx [hunk 16]
```diff
@@ -264,11 +264,22 @@ std::string ROOT::Experimental::RFieldBase::GetQualifiedFieldName() const
...
 
 ROOT::RResult<std::unique_ptr<ROOT::Experimental::RFieldBase>>
-ROOT::Experimental::RFieldBase::Create(const std::string &fieldName, const std::string &typeName)
+ROOT::Experimental::RFieldBase::Create(const std::string &fieldName, const std::string &typeName,
+                                       const RCreateFieldOptions &options)
 {
    auto typeAlias = Internal::GetNormalizedTypeName(typeName);
    auto canonicalType = Internal::GetNormalizedTypeName(GetCanonicalTypeName(typeAlias));
-   return R__FORWARD_RESULT(RFieldBase::Create(fieldName, canonicalType, typeAlias));
+   return R__FORWARD_RESULT(RFieldBase::Create(fieldName, canonicalType, typeAlias, options));
```
Comment (hahnjo) [inline tree/ntuple/v7/src/RFieldBase.cxx]: Should this just forward to `Create(fieldName, typeName, options, nullptr, kInvalidDescriptorId)`?

File: tree/ntuple/v7/test/ntuple_emulated.cxx [hunk 17]
```diff
@@ -93,3 +94,108 @@ struct Outer {
...
+   // To achieve this "forgetting" we use fork() so that the parent process doesn't know about the class.
+
+   FileRaii fileGuard("test_ntuple_emulated_fields_vecs.root");
+
+   ExecInFork([&] {
+      // The child process writes the file and exits, but the file must be preserved to be read by the parent.
+      fileGuard.PreserveFile();
+
+      ASSERT_TRUE(gInterpreter->Declare(R"(
+struct Inner {
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_emulated.cxx:110]: You should uniquely prefix the `struct` names to prevent any caching in the parent process when reading the file.

File: tree/ntuple/v7/inc/ROOT/RFieldBase.hxx [hunk 18]
```diff
@@ -69,15 +69,18 @@ This is and can only be partially enforced through C++.
...
    friend struct ROOT::Experimental::Internal::RFieldCallbackInjector;       // used for unit tests
    friend struct ROOT::Experimental::Internal::RFieldRepresentationModifier; // used for unit tests
    friend void Internal::CallFlushColumnsOnField(RFieldBase &);
    friend void Internal::CallCommitClusterOnField(RFieldBase &);
    friend void Internal::CallConnectPageSinkOnField(RFieldBase &, Internal::RPageSink &, NTupleSize_t);
    friend void Internal::CallConnectPageSourceOnField(RFieldBase &, Internal::RPageSource &);
+   // for RFieldBase::Create
+   friend std::unique_ptr<ROOT::Experimental::RFieldBase>
+   ROOT::Experimental::RFieldDescriptor::CreateField(const RNTupleDescriptor &ntplDesc,
+                                                     const RCreateFieldOptions &options) const;
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/RFieldBase.hxx]: I think this `friend` declaration becomes redundant with the next commit

File: tree/ntuple/v7/test/ntuple_emulated.cxx [hunk 19]
```diff
@@ -199,3 +199,152 @@ struct Outer {
...
+   // Test struct containing an empty struct
+
+   FileRaii fileGuard("test_ntuple_emulated_emptystruct.root");
+
+   ExecInFork([&] {
+      // The child process writes the file and exits, but the file must be preserved to be read by the parent.
+      fileGuard.PreserveFile();
+
+      ASSERT_TRUE(gInterpreter->Declare(R"(
+struct Inner {
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_emulated.cxx:214]: Should also be prefixed, see comment on a previous commit

File: tree/ntuple/v7/src/RNTupleDescriptor.cxx [hunk 20]
```diff
@@ -125,7 +125,7 @@ ROOT::Experimental::RFieldDescriptor::CreateField(const RNTupleDescriptor &ntplD
       // NOTE: Unwrap() here may throw an exception, hence the try block.
       // If options.fReturnInvalidOnError is false we just rethrow it, otherwise we return an InvalidField wrapping the
       // error.
-      auto field = RFieldBase::Create(fieldName, typeName, typeName, options, &ntplDesc, fFieldId).Unwrap();
+      auto field = Internal::CallFieldBaseCreate(fieldName, typeName, typeName, options, &ntplDesc, fFieldId).Unwrap();
```
Comment (hahnjo) [inline tree/ntuple/v7/src/RNTupleDescriptor.cxx]: I'm not sure, but doesn't that mean that the previous commit doesn't compile?

**Review by hahnjo — APPROVED** [review]
LG from my side; there are some stray lines in commit messages that look like the result of squashing, and one comment about a non-compiling commit (not sure)
